### ISSUE TITLE
Port twisted.protocols.tls to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6139]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6139|
|Type|enhancement|
|Created|2012-10-16 18:13:57Z|
|Branch|https://github.com/twisted/twisted/tree/tls-py3-6139|

`twisted.protocols.tls` and `twisted.protocols.test.test_tls` should run on Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__6139 6139
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_tls_py3_6139 branches/tls-py3-6139
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1350411237000000 1350411237000000
changetime__1350999394000000 1350999394000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [36209]) Branching to 'tls-py3-6139'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

(In [36214]) Don't import twisted.internet.ssl until it is ported. Refs [#6139](https://github.com/twisted/twisted/issues/6139)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

(In [36215]) Fix a bug in the loopback port. Refs [#6139](https://github.com/twisted/twisted/issues/6139)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set owner to @exarkun|
|-|-|

  1. I had to manually disable importing of `twisted.internet.ssl` in a couple of test modules (I didn't check if this is because `twisted.protocols.tls` is importable or because I installed pyOpenSSL).
  1. We didn't do the loopback port quite right, so I had to fix a bug there.
  1. I re-implemented a couple of context factories, and skipped the one test that actually requires `_sslverify`. If you want to tell me how to do that easily I can re-enable it as part of this branch.
  1. Other than that, the usual "make the tests pass with extra bees" sort-of-thing.
  1. `twisted.internet.test.test_tls` will be done separately (and in any case probably depends on other unported modules), there's enough bullet points in here as it is.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/tls-py3-6139 is started.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. There are some more unicode strings in `test_identityPumpPolicy` in `test_loopback.py`.
  1. I liked the comments in `admin/_twistedpython3.py`.  Maybe they should annotate the modules being put in `almostModules` instead, though?
  1. I wonder why `twisted/test/test_internet.py` needs changes related to SSL.  The tests should already have skipped based on the reactor not providing `IReactorSSL`.  Are we not properly declaring what interfaces the reactor implements?
  1. There's some missing test coverage for `twisted.protocols.tls`.  Can you file a ticket for adding that test coverage?
  1. `twisted/internet/test/test_tls.py` should have some notes about switching back to the original context factories once they're ported.

As usual, this code doesn't quite fit in my head.  I don't quite understand what the deal with `bio_read` is here, for example.  But the changes don't appear too surprising and the test pass...  Please merge after addressing the above points.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

1.Fixed (r36243)
 1. Fixed, and moved a couple more modules into the `almostModules` list (r36243).
 1. We declare reactor provides `IReactorSSL` if pyopenssl is importable, we haven't really dealt with case where it imports but Twisted's ssl code is broken.
 1. [#6155](https://github.com/twisted/twisted/issues/6155).
 1. Fixed (r36243).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set status to `closed`|
|-|-|

(In [36244]) Merge tls-py3-6139.

Author: itamar
Review: exarkun
Fixes: [#6139](https://github.com/twisted/twisted/issues/6139)

Port twisted.protocols.tls to Python 3.
